File: client/cpp/test/test_basic.cpp


#include "hybridsearch_client.h"
#include "iostream"

int main() {
    hybridsearch::client::Client hybridsearch_client = hybridsearch::client::Client::Connect("127.0.0.1", 23817);

    hybridsearch::client::ClientStatus status = hybridsearch_client.CreateDatabase("db1");
    if(!status.IsOK()) {
        std::cerr << status.ErrorMessage() << std::endl;
        return static_cast<int>(status.ErrorCode());
    }

    std::vector<std::string> db_names = hybridsearch_client.ListDatabases();
    for(auto& db_name: db_names) {
        std::cout << db_name << std::endl;
    }

    status = hybridsearch_client.DropDatabase("db1");
    if(!status.IsOK()) {
        std::cerr << status.ErrorMessage() << std::endl;
        return static_cast<int>(status.ErrorCode());
    }

    db_names = hybridsearch_client.ListDatabases();
    for(auto& db_name: db_names) {
        std::cout << db_name << std::endl;
    }

    hybridsearch_client.Disconnect();
    return 0;
}

File: python/test_cluster/test_basic.py
import time

import pytest
from hybridsearch_cluster import hybridsearchCluster
from mocked_hybridsearch_cluster import MockhybridsearchCluster
import time


def test_admin(cluster: hybridsearchCluster):
    with cluster:
        cluster.add_node("test", "conf/leader.toml")
        hybridsearch1 = cluster.client("test")
        assert hybridsearch1 is not None
        res = hybridsearch1.show_current_node()
        print(res.node_name)
        assert (res.node_role == "admin")


def test_standalone(cluster: hybridsearchCluster):
    with cluster:
        cluster.add_node("test", "conf/pytest_parallel_hybridsearch_conf.toml")
        cluster.set_standalone("test")
        test_client = cluster.client("test")
        assert test_client is not None

        test_client.create_database("db1")
        test_client.drop_database("db1")


def test_0(cluster: hybridsearchCluster):
    with cluster:
        cluster.add_node("node1", "conf/leader.toml")
        cluster.add_node("node2", "conf/follower.toml")

        cluster.set_leader("node1")
        cluster.set_follower("node2")

        time.sleep(1)

        cluster.remove_node("node2")
        cluster.remove_node("node1")


# def test_mock(mock_cluster: MockhybridsearchCluster):
#     cluster = mock_cluster
#     with cluster:
#         cluster.add_node("node1", "conf/leader.toml")
#         cluster.add_node("node2", "conf/follower.toml")

#         cluster.set_leader("node1")
#         cluster.set_follower("node2")

#         time.sleep(1)

#         cluster.disconnect("node2")
#         time.sleep(0.1)
#         cluster.reconnect("node2")

#         cluster.block_peer_net("node2")
#         time.sleep(0.1)
#         cluster.restore_peer_net("node2")

#         time.sleep(1)

#         cluster.remove_node("node2")
#         cluster.remove_node("node1")


# @pytest.mark.docker
# def test_docker(docker_cluster: DockerhybridsearchCluster):
#     cluster = docker_cluster

#     cluster.add_node("node1", "conf/leader.toml")
#     cluster.add_node("node2", "conf/follower.toml")

#     print("init nodes")

#     cluster.set_leader("node1")
#     cluster.set_follower("node2")

#     time.sleep(1)

#     cluster.disconnect("node2")
#     time.sleep(0.1)
#     cluster.reconnect("node2")

#     res = cluster.client("node1").list_databases()
#     print(res.db_names)

#     time.sleep(1)

#     print("remove nodes")

#     cluster.clear()

# # n1 admin, n2 admin, n2 connect n1 as follower-> failed, n2 connect n1 as learner-> failed
# # n1 leader, n2 follower, n3 connect n2 as follower-> failed, n3 connect n2 as learner-> failed
# #
